libplatsupport/plat_include/star64/platsupport/plat/timer.h

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2023, UNSW
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+#pragma once
+
+#include <stdint.h>
+#include <stdbool.h>
+
+/*
+ * The JH7110 SoC contains a timer with four 32-bit counters. Each one of these
+ * counters is referred to as a "channel".
+ */
+
+/* Channel information */
+#define STARFIVE_TIMER_NUM_CHANNELS 4
+#define STARFIVE_TIMER_CHANNEL_REGISTERS_LEN_IN_BYTES 0x40
+#define STARFIVE_TIMER_CHANNEL_0_IRQ 69
+#define STARFIVE_TIMER_CHANNEL_1_IRQ 70
+#define STARFIVE_TIMER_CHANNEL_2_IRQ 71
+#define STARFIVE_TIMER_CHANNEL_3_IRQ 72
+
+/* This information comes from the DTS file */
+#define STARFIVE_TIMER_BASE 0x13050000
+#define STARFIVE_TIMER_REGISTER_WINDOW_LEN_IN_BYTES 0x10000
+#define STARFIVE_TIMER_TICKS_PER_SECOND 0x16e3600
+
+#define STARFIVE_TIMER_MAX_TICKS 0xffffffff
+
+/* Register value constants */
+#define STARFIVE_TIMER_MODE_CONTINUOUS 0
+#define STARFIVE_TIMER_MODE_SINGLE 1
+#define STARFIVE_TIMER_DISABLED 0
+#define STARFIVE_TIMER_ENABLED 1
+#define STARFIVE_TIMER_INTERRUPT_UNMASKED 0
+#define STARFIVE_TIMER_INTERRUPT_MASKED 1
+#define STARFIVE_TIMER_INTCLR_BUSY BIT(1)
+
+typedef struct {
+    /* Registers */
+    /* this register doesn't seem to do anything */
+    uint32_t status;
+    uint32_t ctrl;
+    uint32_t load;
+    uint32_t unknown1;
+    uint32_t enable;
+    uint32_t reload;
+    uint32_t value;
+    uint32_t unknown2;
+    uint32_t intclr;
+    uint32_t intmask;
+} starfive_timer_regs_t;
+
+typedef struct {
+    volatile starfive_timer_regs_t *regs;
+    /*
+     * Stores the number of times the continuous counter timer has elapsed and started over.
+     * This allows us to count to a higher number than allowed by the hardware.
+     */
+    uint32_t value_h;
```

Comment:
Should this be declared volatile also, as it could be access concurrently, and we also want to prevent the compiler from doing some optimization that would move the accesses around.

Reply:
I'm confused, these drivers aren't multithreaded right? How could there be concurrent access?

Reply:
Driver ware not multi-threaded, but they have an ISR that could interrupt us any time and update this value. Using `volatile` might not prevent the race completely, but could makes the windows even smaller.